### ISSUE TITLE
strange issue using pipes and qplot

### ISSUE THREAD

**Issue by casallas** [smbache/magrittr#62]
All of the following calls
```r
rnorm(1000) %>% qplot
rnorm(1000) %>% qplot(.)
rnorm(1000) %>% {qplot(.)}
```
Yield the following error
```
Error in as.data.frame.default(x[[i]], optional = TRUE, stringsAsFactors = stringsAsFactors) : 
  cannot coerce class "c("gg", "ggplot")" to a data.frame
```
Strangely, however, the following works
```r
rnorm(1000) %>% {x=.; qplot(x)}
```
but that breaks the pipe magic... something else that works is `debug(qplot)`, then calling `str(x)` once debugging, which makes me think there may be some sort of lazy evaluation happening (perhaps in `ggplot2`?) which doesn't play nicely with the `.` placeholder?

I can confirm that this wasn't an issue in magrittr 1.1.0

**Comment by gaborcsardi** [comment]
[This was closed some time ago, but my question is really related to the recursive version.]

Do you have an estimate or just a guess what this means for memory allocation along the pipe? 

To elaborate, without promises, the recursive version allocates all `value`s along the pipe, *at the same time*.

Now, we of course have promises, so quite often this does not happen, and simply a promise is pushed through the pipe, and `value` is only evaluated at the last stage.

With the for version, there are no promises, but there is no nesting either, so no memory accumulation along the pipe.

**Comment by smbache** [comment]
No I'm not sure how costly it is, although surely recursion is inefficient, but this is a problem linear in the number of pipes (and almost always negligible). However there were examples where the loop version failed for subtle reasons. I'm sure @kmillar and crew gets the recursion overhead down ;)

Did you experience trouble with the recursive version?

**Comment by gaborcsardi** [comment]
@smbache No, I haven't seen trouble, so, yes, this is theoretical.

I am not worried about the recursion per se, that's fine with me. I am "worried" about having a `value` at each level of the recursion. 

Think about a typical use case of putting a (largish) data frame into a pipe, and manipulating and transforming it at each step. If you write it without pipes:

```r
result <- manipulate1(DF)
result <- manipulate2(result)
result <- manipulate3(result)
...
```

then there is only one copy of `result`. (Well, usually two really, but the point is that it is O(1).)

However if you write it with pipes, then, because of the recursion, there will be a separate `value` at each level of the recursion. (AFAIK there is no tail-recursion optimization in R.)
If these `value`s are large data frames, then you'll end up with multiple copies of them. I.e. if you have a pipe with 10 stages, you store the "data" 10 times, **at the same time**.

Of course this is usually **not true** because of the promises. If there are no side effects, R is just passing `value` as a promise down the pipe, and everything works effectively the same way as the non-pipey version.

So my question is: how often can we push a promise to the end of the pipe?

In other words, I am fairly sure that by introducing side effects, I can make R store the data as many times as many stages I have in the pipe, so I can "break" it. But does it really happen in real uses cases? Do people have side effects in the pipe stages?

I guess it might be also a good idea to promote pipey code without side effects. E.g. if the manual, readmes, vignettes, etc. have code with side effects, maybe we can reconsider them.
